spark-cassandra-connector/src/main/scala/org/apache/spark/sql/cassandra/CassandraSourceRelation.scala

```diff
@@ -0,0 +1,232 @@
+}
+
+/** Table scan relation implements [[BaseRelation]], [[InsertableRelation]] and [[TableScan]] */
+private[cassandra] class TableScanRelationImpl(table: String, keyspace: String, cluster: Option[String],
+                                               connector: CassandraConnector, readConf: ReadConf, writeConf: WriteConf,
+                                               userSpecifiedSchema: Option[StructType],
+                                               override val sqlContext: SQLContext)
+  extends BaseRelationImpl(table, keyspace, cluster, connector, readConf, writeConf, userSpecifiedSchema,
+    sqlContext) with TableScan {
+  override def buildScan(): RDD[Row] = super.buildScan()
+}
+
+/** Table scan relation implements [[BaseRelation]], [[InsertableRelation]] and [[PrunedScan]] */
+private[cassandra] class PrunedScanRelationImpl(table: String, keyspace: String, cluster: Option[String],
+                                                connector: CassandraConnector, readConf: ReadConf, writeConf: WriteConf,
+                                                userSpecifiedSchema: Option[StructType],
+                                                override val sqlContext: SQLContext)
+  extends BaseRelationImpl(table, keyspace, cluster, connector, readConf, writeConf, userSpecifiedSchema, sqlContext)
+  with PrunedScan {
+  override def buildScan(requiredColumns: Array[String]): RDD[Row] = {
+    val transformer = new RddFilterSelectPdTrf(requiredColumns, columnNameByLowercase, Seq.empty)
+    transformer.maybeSelect apply baseRdd
+  }
+}
+
+/** Table scan relation implements [[BaseRelation]], [[InsertableRelation]] and [[PrunedFilteredScan]] */
+private[cassandra] class PrunedFilteredScanRelationImpl(table: String, keyspace: String, cluster: Option[String],
+                                                        connector: CassandraConnector, readConf: ReadConf,
+                                                        writeConf: WriteConf,
+                                                        userSpecifiedSchema: Option[StructType],
+                                                        override val sqlContext: SQLContext)
+  extends BaseRelationImpl(table, keyspace, cluster, connector, readConf, writeConf, userSpecifiedSchema, sqlContext)
+  with PrunedFilteredScan {
+
+  override def buildScan(requiredColumns: Array[String], filters: Array[Filter]): RDD[Row] = {
+    val pushDown = new FilterPushDown(filters, tableDef)
+    val pushdownFilters = pushDown.toPushDown
+    val preservedFilters = pushDown.toPreserve
+
+    logInfo(s"pushdown filters: ${pushdownFilters.toString()}")
+
+    val dataTypeMapping: Map[String, (Int, DataType)] = requiredColumns.map(column =>
+      (column, (requiredColumns.indexOf(column),
+        ColumnDataType.catalystDataType(tableDef.columnByName.get(column).
+          getOrElse(throw new RuntimeException(s"Can't find column $column in $table")).columnType,
```

Comment:
When you need to call `getOrElse` only to throw an Exception on "else", then you might wish not to use `get` / `getOrElse` at all, but just use plain `columnByName(column)`. It will throw NoSuchElementException by default. To make the error message better, you can amend the `columnByName` map `withDefault(key => throw ...)`. Then you don't need to repeat the same message in multiple places and error handling is not mixed up with code.